README.md
# Length
[![Discord](https://img.shields.io/discord/798936842808197140?logo=discord)](https://discord.gg/st6SqenRnm)  
A programming language based on line length
## Guide
| Line Length | Name    | Description                                                                      |
|-------------|---------|----------------------------------------------------------------------------------|
| 9           | `inp`   | Pushes the ascii value of the first byte of stdin to the stack.                  |
| 10          | `add`   | Adds the top two values on the stack and pushes the result onto the stack.       |
| 11          | `sub`   | Subtracts the top two values on the stack and pushes the result onto the stack.  |
| 12          | `dup`   | Duplicates the top value of the stack.                                           |
| 13          | `cond`  | If the top value of the stack is 0, skip the next instruction. Then pop it.      |
| 14          | `gotou` | Sets the program counter to the value of the line under the instruction.         |
| 15          | `outn`  | Pops the top of the stack, and outputs it as a number.                           |
| 16          | `outa`  | Pops the top of the stack, and outputs its ascii value.                          |
| 17          | `rol`   | Rotates the stack to the left: `[ 7 6 5 ]` -> `[ 6 5 7 ]`                        |
| 18          | `swap`  | Swaps the top two values of the stack: `[ 7 6 5 ]` -> `[ 6 7 5 ]`                |
| 20          | `mul`   | Multiplies the top two values on the stack and pushes the result onto the stack. |
| 21          | `div`   | Divides the top two values on the stack and pushes the result onto the stack.    |
| 23          | `pop`   | Pops the top value of the stack                                                  |
| 24          | `gotos` | Sets the program counter to the value on top of the stack                        |
| 25          | `push`  | Pushes the value of the line under it to the stack, and skips that line          |
| 27          | `ror`   | Rotates the stack to the right: `[ 7 6 5 ]` -> `[ 5 7 6 ]`                       |

Esolangs link: https://esolangs.org/wiki/Length (go here instead, its a bit more detailed)


CHANGELOG.md
## 0.1.0

- Initial version, created by Stagehand

## 0.2.0

- Add a CLI, rename to Length, and allow user to toggle debug mode

## 0.2.1

- Improve description, add examples

## 0.3.0

- pop, rotation, and swapping instructions

lib/lines.dart
import 'dart:io';

import 'package:stack/stack.dart';

const DEBUG = false;

/// The interpreter for the Lines esolang
class Lines {
  /// Contains the program as a list of line lengths
  List<int> _program;

  /// Contains the stack. The only data structure in Lines is an int
  Stack<int> _stack;

  /// Loads a program from a file
  Lines(File program) {
    _program = [];
    _stack = Stack();
    for (var line in program.readAsLinesSync()) {
      _program.add(line.length);
    }
  }

  /// Runs the program
  void run() {
    for (var i = 0; i < _program.length; i++) {
      var instruction = _program[i];
      if (DEBUG) {
        stdout.write('(${i + 1})');
      }
      switch (instruction) {
        // addition
        case 10:
          if (DEBUG) {
            print('add');
          }
          if (_stack.length < 2) {
            throw ArgumentError('Stack underflow at line ${i}');
          }

          var num1 = _stack.pop();
          var num2 = _stack.pop();
          _stack.push(num1 + num2);
          break;
        // subtraction
        case 11:
          if (DEBUG) {
            print('sub');
          }
          if (_stack.length < 2) {
            throw ArgumentError('Stack underflow at line ${i}');
          }

          var num1 = _stack.pop();
          var num2 = _stack.pop();
          _stack.push(num2 - num1);
          break;
        // multiplication
        case 20:
          if (DEBUG) {
            print('mul');
          }
          if (_stack.length < 2) {
            throw ArgumentError('Stack underflow at line ${i}');
          }

          var num1 = _stack.pop();
          var num2 = _stack.pop();
          _stack.push(num1 * num2);
          break;
        // division
        case 21:
          if (DEBUG) {
            print('div');
          }
          if (_stack.length < 2) {
            throw ArgumentError('Stack underflow at line ${i}');
          }
          var num1 = _stack.pop();
          var num2 = _stack.pop();
          _stack.push(num2 / num1 as int);
          break;
        // Output as number
        case 15:
          if (DEBUG) {
            print('outn::::::::::');
          }
          if (_stack.length < 1) {
            throw ArgumentError('Stack underflow at line ${i}');
          }
          stdout.write(_stack.pop());
          break;
        // Output as ASCII
        case 16:
          if (DEBUG) {
            print('outa:::::::::');
          }
          if (_stack.length < 1) {
            throw ArgumentError('Stack underflow at line ${i}');
          }
          stdout.write(String.fromCharCode(_stack.pop()));
          break;
        // Push
        case 25:
          if (DEBUG) {
            print('push');
          }
          _stack.push(_program[++i]);
          break;
        // conditional
        case 13:
          if (DEBUG) {
            print('cond');
          }
          if (_stack.length < 1) {
            throw ArgumentError('Stack underflow at line ${i}');
          }
          if (_stack.pop() == 0) {
            i++;
            if ([14, 25].contains(_program[i])) {
              i++;
            }
          }
          break;
        // Goto (Line under)
        case 14:
          if (DEBUG) {
            print('gotou');
          }
          i = _program[i + 1] - 1;
          if (DEBUG) {
            print('pc = ${i}');
          }
          break;
        // Goto (Top of stack)
        case 24:
          if (DEBUG) {
            print('gotos');
          }
          if (_stack.length < 1) {
            throw ArgumentError('Stack underflow at line ${i}');
          }
          i = _stack.pop();
          if (DEBUG) {
            print('pc = ${i}');
          }
          break;
        // Input
        case 9:
          print('Input:');
          _stack.push(stdin.readByteSync());
          break;
        // dup
        case 12:
          if (DEBUG) {
            print('dup');
          }
          var top = _stack.pop();
          _stack.push(top);
          _stack.push(top);
      }
      if (DEBUG) {
        print('----------');
        _stack.print();
      }
    }
  }
}


lib/stack.dart
import 'dart:collection';
import 'dart:core' as core;
import 'dart:core';

import 'dart:io';

class Stack<T> {
  final ListQueue<T> list = ListQueue();

  final int noLimit = -1;

  /// the maximum number of entries allowed on the stack. -1 = no limit.
  int _sizeMax = 0;

  /// Default constructor sets the maximum stack size to 'no limit.'
  Stack() {
    _sizeMax = noLimit;
  }

  /// Constructor in which you can specify maximum number of entries.
  /// This maximum is a limit that is enforced as entries are pushed on to the stack
  /// to prevent stack growth beyond a maximum size. There is no pre-allocation of
  /// slots for entries at any time in this library.
  Stack.sized(int sizeMax) {
    if (sizeMax < 2) {
      throw core.Exception('Will never happen');
    } else {
      _sizeMax = sizeMax;
    }
  }

  /// check if the stack is empty.
  bool get isEmpty => list.isEmpty;

  /// check if the stack is not empty.
  bool get isNotEmpty => list.isNotEmpty;

  /// push element in top of the stack.
  void push(T e) {
    if (_sizeMax == noLimit || list.length < _sizeMax) {
      list.addLast(e);
    } else {
      throw core.Exception('Will never happen');
    }
  }

  /// get the top of the stack and delete it.
  T pop() {
    if (isEmpty) {
      throw core.Exception('Will never happen');
    }
    var res = list.last;
    list.removeLast();
    return res;
  }

  /// get the top of the stack without deleting it.
  T top() {
    if (isEmpty) {
      throw core.Exception('Will never happen');
    }
    return list.last;
  }

  /// get the size of the stack.
  int size() {
    return list.length;
  }

  /// get the length of the stack.
  int get length => size();

  /// returns true if element is found in the stack
  bool contains(T x) {
    return list.contains(x);
  }

  /// print stack
  void print() {
    stdout.write('[ ');
    for (var item in List<T>.from(list).reversed) {
      stdout.write(item);
      stdout.write(' ');
    }
    core.print(']');
  }
}


lib/length.dart
import 'dart:io';

import './stack.dart';

/// The interpreter for the Lines esolang
class Length {
  /// Contains the program as a list of line lengths
  List<int> _program;

  /// Contains the stack. The only data structure in Lines is an int
  Stack<int> _stack;

  /// Loads a program from a file
  Length(File program) {
    _program = [];
    _stack = Stack();
    for (var line in program.readAsLinesSync()) {
      _program.add(line.length);
    }
  }

  /// Runs the program
  void run(bool debug) {
    for (var i = 0; i < _program.length; i++) {
      var instruction = _program[i];
      if (debug) {
        stdout.write('(${i + 1})');
      }
      switch (instruction) {
        // Input
        case 9:
          print('Input:');
          _stack.push(stdin.readByteSync());
          break;
        // addition
        case 10:
          if (debug) {
            print('add');
          }
          if (_stack.length < 2) {
            throw ArgumentError('Stack underflow at line ${i}');
          }

          var num1 = _stack.pop();
          var num2 = _stack.pop();
          _stack.push(num1 + num2);
          break;
        // subtraction
        case 11:
          if (debug) {
            print('sub');
          }
          if (_stack.length < 2) {
            throw ArgumentError('Stack underflow at line ${i}');
          }

          var num1 = _stack.pop();
          var num2 = _stack.pop();
          _stack.push(num2 - num1);
          break;
        // dup
        case 12:
          if (debug) {
            print('dup');
          }
          var top = _stack.pop();
          _stack.push(top);
          _stack.push(top);
          break;
        // conditional
        case 13:
          if (debug) {
            print('cond');
          }
          if (_stack.length < 1) {
            throw ArgumentError('Stack underflow at line ${i}');
          }
          if (_stack.pop() == 0) {
            i++;
            if ([14, 25].contains(_program[i])) {
              i++;
            }
          }
          break;
        // Goto (Line under)
        case 14:
          if (debug) {
            print('gotou');
          }
          i = _program[i + 1] - 1;
          if (debug) {
            print('pc = ${i}');
          }
          break;
        // Output as number
        case 15:
          if (debug) {
            print('outn::::::::::');
          }
          if (_stack.length < 1) {
            throw ArgumentError('Stack underflow at line ${i}');
          }
          stdout.write(_stack.pop());
          break;
        // Output as ASCII
        case 16:
          if (debug) {
            print('outa:::::::::');
          }
          if (_stack.length < 1) {
            throw ArgumentError('Stack underflow at line ${i}');
          }
          stdout.write(String.fromCharCode(_stack.pop()));
          break;
        // rotate left
        case 17:
          if (debug) {
            print('rol');
          }
          if (_stack.length < 1) {
            break;
          }
          var first = _stack.pop();
          _stack.list.addFirst(first);
          break;
        // swap
        case 18:
          if (debug) {
            print('swap');
          }
          if (_stack.length < 2) {
            throw ArgumentError('Stack underflow at line ${i}');
          }
          var a = _stack.pop();
          var b = _stack.pop();

          _stack.push(a);
          _stack.push(b);
          break;
        // multiplication
        case 20:
          if (debug) {
            print('mul');
          }
          if (_stack.length < 2) {
            throw ArgumentError('Stack underflow at line ${i}');
          }

          var num1 = _stack.pop();
          var num2 = _stack.pop();
          _stack.push(num1 * num2);
          break;
        // division
        case 21:
          if (debug) {
            print('div');
          }
          if (_stack.length < 2) {
            throw ArgumentError('Stack underflow at line ${i}');
          }
          var num1 = _stack.pop();
          var num2 = _stack.pop();
          _stack.push(num2 / num1 as int);
          break;
        // swap
        case 23:
          if (debug) {
            print('pop');
          }
          if (_stack.length < 1) {
            throw ArgumentError('Stack underflow at line ${i}');
          }
          _stack.pop();
          break;
        // Goto (Top of stack)
        case 24:
          if (debug) {
            print('gotos');
          }
          if (_stack.length < 1) {
            throw ArgumentError('Stack underflow at line ${i}');
          }
          i = _stack.pop();
          if (debug) {
            print('pc = ${i}');
          }
          break;
        // Push
        case 25:
          if (debug) {
            print('push');
          }
          _stack.push(_program[++i]);
          break;

        // rotate right
        case 27:
          if (debug) {
            print('ror');
          }
          if (_stack.length < 1) {
            break;
          }
          var first = _stack.list.removeFirst();
          _stack.push(first);
          break;
      }
      if (debug) {
        print('----------');
        _stack.print();
      }
    }
  }
}
